ocaml/testsuite/tests/typing-layouts/rep_poly.ml

```diff
@@ -0,0 +1,475 @@
+         end
+       is not included in
+         sig external id : 'a -> 'a = "%identity" end
+       Values do not match:
+         external id : ('a : any). 'a -> 'a = "%identity" [@@rep_poly]
+       is not included in
+         external id : 'a -> 'a = "%identity"
+       The two primitives have different [@rep_poly] attributes
+|}]
+
+module S : sig
+  external[@rep_poly] id : ('a : any). 'a -> 'a = "%identity"
+end = struct
+  external id : ('a : any). 'a -> 'a = "%identity"
+end
+
+[%%expect{|
+Lines 3-5, characters 6-3:
+3 | ......struct
+4 |   external id : ('a : any). 'a -> 'a = "%identity"
+5 | end
+Error: Signature mismatch:
+       Modules do not match:
+         sig external id : 'a -> 'a = "%identity" end
+       is not included in
+         sig
+           external id : ('a : any). 'a -> 'a = "%identity" [@@rep_poly]
+         end
+       Values do not match:
+         external id : 'a -> 'a = "%identity"
+       is not included in
+         external id : ('a : any). 'a -> 'a = "%identity" [@@rep_poly]
+       The two primitives have different [@rep_poly] attributes
+|}]
+
+(* External in struct *)
+
+module S : sig
+  val id : ('a : float64). 'a -> 'a
+end = struct
+  external[@rep_poly] id : ('a : any). 'a -> 'a = "%identity"
+end
+
+let () = Format.printf "%f\n" (F.to_float (S.id #1.)); Format.print_flush ()
+
+[%%expect{|
+module S : sig val id : ('a : float64). 'a -> 'a end
+1.000000
+|}]
+
+let () = Format.printf "%s\n" (S.id "abc")
+
+[%%expect{|
+Line 1, characters 36-41:
+1 | let () = Format.printf "%s\n" (S.id "abc")
+                                        ^^^^^
+Error: This expression has type string but an expression was expected of type
+         ('a : float64)
+       The layout of string is value, because
+         it is the primitive value type string.
+       But the layout of string must be a sublayout of float64, because
+         of the definition of id at line 2, characters 2-35.
+|}]
+
+
+module S : sig
+  val id : ('a : any). 'a -> 'a
+end = struct
+  external[@rep_poly] id : ('a : any). 'a -> 'a = "%identity"
+end
+
+[%%expect{|
+Lines 3-5, characters 6-3:
+3 | ......struct
+4 |   external[@rep_poly] id : ('a : any). 'a -> 'a = "%identity"
+5 | end
+Error: Signature mismatch:
+       Modules do not match:
+         sig
+           external id : ('a : any). 'a -> 'a = "%identity" [@@rep_poly]
+         end
+       is not included in
+         sig val id : ('a : any). 'a -> 'a end
+       Values do not match:
+         external id : ('a : any). 'a -> 'a = "%identity" [@@rep_poly]
+       is not included in
+         val id : ('a : any). 'a -> 'a
+       The type 'a -> 'a is not compatible with the type 'b -> 'b
+       The layout of 'a is any, because
+         of the definition of id at line 2, characters 2-31.
+       But the layout of 'a must be representable, because
+         it's the representation polymorphic type in an external declaration.
```

Comment:
I agree the message here is not ideal but it's also a bit tricky to fix.

> I think it's the case that a rep_poly external in a struct must always be a rep_poly external in the sig, so we could say explicitly that the signature must have that shape.

This is currently not true because we allow programs like:

```ocaml
module S : sig
  val id : ('a : float64). 'a -> 'a
end = struct
  external[@rep_poly] id : ('a : any). 'a -> 'a = "%identity"
end
```

And it's hard to tell if the jkind error is the result of using `[@rep_poly]` or not without exposing the layout history information in some way. The program below produces a similar error where we want to keep the layout history:

```ocaml
module S : sig
  val id : ('a : float64). 'a -> 'a
end = struct
  external[@rep_poly] id: ('a : bits32). 'a -> 'a = "%identity"
end
```

It might make more sense to educate the programmers to read `('a : any)` in an `external[@rep_poly]` as being representable (which is what it it) instead of maintaining the illusion that it has jkind any.

This relates to the comment below where we would have to say "Argument and return types of external definitions must have a representable layout unless the external is explicitly marked representation polymorphic" (which is also not quite true since you can't have a `t_any` even if you use `[@rep_poly]`). If we are explicit in stating that type variables with jkind any in a `[@rep_poly]` are special and have representable layouts, then an error message like "types in an external must have a representable layout" would suffice.

What do you think about this? I agree it can get confusing.

Reply:
At some point, we'll have composite layouts (instead of just single words), and we'll probably want extension fields there eventually, So we could have `id : ('a : [%layout_poly]). 'a -> 'a` or something. That would be closer to the truth than what we have now.

But I think I favor Alan's proposed solution, which is to educate users. If an `external[@layout_poly]` is involved, can we append a little text to the message reminding users that `[@layout_poly] forces all variables of layout 'any' to be representable at call sites` or something?

Reply:
We can definitely have that in the layout history message. And perhaps the `types in an external must have a representable layout` message would be another place to add it

Reply:
Added the text hint to both the layout history message and the `types in an external ...` error